substra/sdk/compute_plan.py

```diff
@@ -0,0 +1,157 @@
+            assert not compositetuple.out_trunk_model_permissions.public
+            _insert_into_graph(
+                tuple_graph=tuple_graph,
+                tuple_id=compositetuple.composite_traintuple_id,
+                in_model_ids=[
+                    model
+                    for model in [
+                        compositetuple.in_head_model_id,
+                        compositetuple.in_trunk_model_id,
+                    ]
+                    if model
+                ],
+            )
+            compositetuples[compositetuple.composite_traintuple_id] = compositetuple
+
+    return tuple_graph, traintuples, aggregatetuples, compositetuples
+
+
+def auto_batching(spec, is_creation: bool = True, batch_size: int = 20):
+    """Auto batching of the compute plan tuples
+    """
+
+    # Create the dependency graph and get the dict
+    # of tuples by id
+    (
+        tuple_graph,
+        traintuples,
+        aggregatetuples,
+        composite_traintuples,
+    ) = get_dependency_graph(spec)
+
+    already_created_keys = set()
+    if not is_creation:
+        # Here we get the pre-existing tuples and assign them the minimal rank
+        for dependencies in tuple_graph.values():
+            for dependency_key in dependencies:
+                if dependency_key not in tuple_graph:
+                    already_created_keys.add(dependency_key)
+
+    # Compute the relative ranks of the new tuples (relatively to each other, these
+    # are not their actual ranks in the compute plan)
+    visited = graph.compute_ranks(
+        node_graph=tuple_graph, node_to_ignore=already_created_keys
+    )
+
+    # Add the testtuples to 'visited' to take them into account in the batches
+    testtuples = dict()
+    if spec.testtuples:
+        for testtuple in spec.testtuples:
+            visited["test_" + testtuple.traintuple_id] = (
+                visited[testtuple.traintuple_id] + 1
+            )
+            testtuples["test_" + testtuple.traintuple_id] = testtuple
+
+    # Sort the tuples by rank
+    sorted_by_rank = sorted(visited.items(), key=lambda item: item[1])
+
+    # Create / update by batch
+    for i in range(math.ceil(len(sorted_by_rank) / batch_size)):
+        start = i * batch_size
+        end = min(len(sorted_by_rank), (i + 1) * batch_size)
+
+        if i == 0 and is_creation:
+            # Compute plan does not exist (ie first batch of a creation):
+            # we create it
+            tmp_spec = deepcopy(spec)
+            (
+                tmp_spec.traintuples,
+                tmp_spec.aggregatetuples,
+                tmp_spec.composite_traintuples,
+                tmp_spec.testtuples,
+            ) = graph.filter_tuples_in_list(
+                sorted_by_rank[start:end],
+                traintuples,
+                aggregatetuples,
+                composite_traintuples,
+                testtuples,
+            )
+            yield tmp_spec
```

Comment:
A generator, nice!

Reply:
I try to take advantage of them more, it seemed like a good occasion !